Parallelize OPTIMIZE FINAL for non-replicated table

aiven-sal opened the pull request:
Close #46770

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Performance Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
On  non-replicated tables, OPTIMIZE FINAL schedules merges for all partitions and waits for them to complete

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

aiven-sal committed:
Parallelize OPTIMIZE FINAL for non-replicated table

Close #46770

robot-ch-test-poll commented:
<!-- automatic status comment for PR #49504 from aiven-sal/ClickHouse:aiven-sal/optimize_final -->
This is an automated comment for commit 831da0b6119cfdeb8812a0e38e2015090fb3f1db with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49504/831da0b6119cfdeb8812a0e38e2015090fb3f1db/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

aiven-sal committed:
Wait for completion of background OPTIMIZE FINAL merges

Existing tests expect OPTIMIZE FINAL to only return after
the merges are completed.

aiven-sal committed:
Fix tests

When OPTIMIZE FINAL merges happen in background their work is not
accounted for in the query_log's ProfileEvents

al13n321 commented on tests/queries/0_stateless/02293_selected_rows_and_merges.sh:
```diff
@@ -15,7 +15,7 @@ ${CLICKHOUSE_CLIENT} -q "insert into tt select number, 1 from numbers(1e6)"
 ${CLICKHOUSE_CLIENT} --optimize_throw_if_noop 1 -q "optimize table tt final" "--query_id=$query_id"
 
-# Here SelectRows and SelectBytes should be zero, MergedRows is 2m and MergedUncompressedBytes is 16m
+# Here for OPTIMIZE FINAL all values are 0, cause merges are executed in background.
```
The query still waits for the merges to complete, it just doesn't propagate the metrics to query_log. A whole new keyword wouldn't make sense for that.

We really should try to propagate the metrics, but I'm not sure what's a good way to do it.

I guess we should pass the thread group from `CurrentThread::getGroup()` in `merge()` to `MergePlainMergeTreeTask`, then make `transfer_profile_counters_to_initial_query()` add metrics to that thread group's `performance_counters`; either from `task_thread_group->performance_counters` (which I suspect won't work, but worth trying) or from `profile_counters` (like in `write_part_log`). I'm not sure if this should replace the `query_thread_group` or be a different thing.

@CheSema, any pointers?

CheSema replied:
That changes will break stats on the query.

Stats like this
https://github.com/ClickHouse/ClickHouse/blob/e708f0d9413a71cb788d3d23731fe379f44fd729/tests/queries/0_stateless/02293_selected_rows_and_merges.sh#L16-L20

@al13n321 is right, something has to be done to preserve it.

The most simple way I think is pass `transfer_profile_counters_to_initial_query calback` outside MergePlainMergeTreeTask. It has to add the stats to the query's thread group.

Be aware that this profile_counters does not account all metrics, the comment is actual.
https://github.com/ClickHouse/ClickHouse/blob/e708f0d9413a71cb788d3d23731fe379f44fd729/src/Storages/MergeTree/MergePlainMergeTreeTask.cpp#L34-36

The better approach is to account stats directly to the query's thread group. But I'm not sure that is safe to just reuse its thread group as a thread group for merge. That group consist query context, which has to have limited influence on merging. So for me this is an actual question how to do it elegant.

CheSema replied:
When `executeHere(task)` called thread group is not shared.

al13n321 replied:
Ah, so `executeHere()` also creates a new thread group, I see.

al13n321 reviewed (changes requested):
IIUC, this makes the merge happen in context of a background thread rather than in query context. So, metrics aren't attributed to the query, memory limits are not taken from the query/session/user settings, and log messages won't include the query. Not ideal.

Let's at least propagate the metrics, then it'll be acceptable.

CheSema reviewed (changes requested):
Query stats have to be preserved.

al13n321 commented:
Ok, I think this is how exactly to make the profile events propagate correctly:
```
src/Storages/MergeTree/MergePlainMergeTreeTask.h:
@@ -49,6 +49,11 @@ public:
         txn = std::move(txn_);
     }
 
+    void setQueryThreadGroup(ThreadGroupPtr g)
+    {
+        query_thread_group = g;
+    }
+
 private:
     void prepare();
     void finish();
@@ -89,6 +94,7 @@ private:
     MergeTreeTransactionPtr txn;
 
     ProfileEvents::Counters profile_counters;
+    ThreadGroupPtr query_thread_group;
 
     ContextMutablePtr task_context;
```
```
src/Storages/MergeTree/MergePlainMergeTreeTask.cpp:
@@ -108,7 +108,10 @@ void MergePlainMergeTreeTask::prepare()
             std::move(profile_counters_snapshot));
     };
 
-    transfer_profile_counters_to_initial_query = [this, query_thread_group = CurrentThread::getGroup()] ()
+    if (!query_thread_group)
+        query_thread_group = CurrentThread::getGroup();
+
+    transfer_profile_counters_to_initial_query = [this] ()
     {
         if (query_thread_group)
         {
```
Then call `task->setQueryThreadGroup(CurrentThread::getGroup())` in `StorageMergeTree::merge()` in the async case.

Please try it and see if it fixes the two tests.

aiven-sal committed:
Add metrics support

Co-authored-by: Michael Kolupaev <michael.kolupaev@clickhouse.com>

aiven-sal committed:
Revert "Fix tests"

This reverts commit ea2e53f58b8229a8f69d8e84792c4bec9b839bae.

al13n321 commented:
Test failures look related, need to investigate: `test_merge_canceled_by_s3_errors`, `test_memory_limit_success`, `01293_optimize_final_force`, `01825_replacing_vertical_merge`.

alexey-milovidov commented:
We have had many optimizations for FINAL recently - it will be nice to also have this one finished.

alexey-milovidov commented:
@aiven-sal Looks abandoned.
